False Positive Error with 'this' and object literal shorthand

**What version of ESLint are you using?**

2.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**

Node v5.7.1, server side code, not using Babel

**Please show your full configuration:**

```
// Based on the awesome config from walmart labs:
//https://github.com/walmartlabs/eslint-config-defaults
{
	"extends": "defaults/configurations/google",
	"rules": {
		"max-len": [2, 110, 2],
		"indent": [2, "tab"],
		"dot-location": [2, "property"],
		"no-mixed-spaces-and-tabs": [2, "smart-tabs"],
		"no-console": [2],
		"no-var": [2]
	},
	"env": {
		"node": true,
		"es6": true,
		"mocha": true
	},
    "ecmaFeatures": {
        "modules": true,
        "classes": true,
        "arrowFunctions": true,
        "blockBindings": true,
        "templateStrings": true
    }
}
```

**What did you do? Please include the actual source code causing the issue.**

```
    myFunc() {

		let that = 1;

		log.debug({
			that,
			this
		}, 'log message here');
    }
```

**What did you expect to happen?**

No error should be returned

**What actually happened? Please include the actual, raw output from ESLint.**

```
  52:4  error  'this' is not defined  no-undef
```

This code works as expected:

```
    myFunc() {

		let that = 1;

		log.debug({
			that,
			this: this
		}, 'log message here');
    }
```

Probably a rare usecase, but in this case we wanted to log 'this' in our logger and have it appear as 'this'.

I think this is a regression, as I _think_ this code has been in place for a while without throwing an error and with the same eslint config. We've just upgraded eslint.

--- a/package.json
+++ b/package.json
@@ -42,7 +42,7 @@
     "doctrine": "^1.2.2",
     "es6-map": "^0.1.3",
     "escope": "^3.6.0",
-    "espree": "3.1.4",
+    "espree": "^3.1.6",
     "estraverse": "^4.2.0",
     "esutils": "^2.0.2",
     "file-entry-cache": "^1.1.1",

--- a/tests/lib/eslint.js
+++ b/tests/lib/eslint.js
@@ -3685,6 +3685,13 @@ describe("eslint", function() {
             eslint.verify("var { a='a' } = {};", { parserOptions: { ecmaVersion: 6 }});
         });
 
+        it("should report syntax error when a keyword exists in object property shorthand", function() {
+            var messages = eslint.verify("let a = {this}", { parserOptions: { ecmaVersion: 6 }});
+
+            assert.equal(messages.length, 1);
+            assert.equal(messages[0].fatal, true);
+        });
+
         it("should not rewrite env setting in core (https://github.com/eslint/eslint/issues/4814)", function() {
 
             // This test focuses on the instance of https://github.com/eslint/eslint/blob/v2.0.0-alpha-2/conf/environments.js#L26-L28

